tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+	ErrCompactionBlock = errors.New("compaction description")
+
+	// within a segment file, it could be a compaction file, which has parts in it that
+	// indicate a file name that was compacted. We use this sequence to identify those sections
+	CompactSequence = []byte{0xFF, 0xFF}
+)
+
+type Log struct {
+	path           string
+	walSegmentSize int
+	flushSize      int
+
+	flush           chan int    // signals a background flush on the given partition
+	flushCheckTimer *time.Timer // check this often to see if a background flush should happen
+
+	// These coordinate closing and waiting for running goroutines.
+	wg      sync.WaitGroup
+	closing chan struct{}
+
+	// Used for out-of-band error messages.
+	logger *log.Logger
+
+	mu         sync.RWMutex
+	partitions map[uint8]*partition
+
+	// The writer used by the logger.
+	LogOutput     io.Writer
+	FlushInterval time.Duration
+	SegmentSize   int
+
+	// Settings that control when a partition should get flushed to index and compacted
+	// flush if any series in the partition has exceeded this size threshold
+	MaxSeriesSize int
+	// a series is ready to flush once it has this much data in it
+	ReadySeriesSize int
+	// a partition is ready to flush if this percentage of series has hit the readySeriesSize or greater
+	CompactionThreshold float64
+	// a partition should be flushed if it has reached this size in memory
+	PartitionSizeThreshold uint64
+	// the number of separate partitions to create for the WAL. Compactions happen per partition. So this number
+	// will affect what percentage of the WAL gets compacted at a time. For instance, a setting of 10 means
+	// we generally will be compacting about 10% of the WAL at a time.
+	PartitionCount uint64
+
+	IndexWriter indexWriter
+}
+
+// indexWriter is an interface for the indexed database the WAL flushes data to
+type indexWriter interface {
+	// time ascending points where each byte array is:
+	//   int64 time
+	//   data
+	WriteIndex(pointsByKey map[string][][]byte) error
+}
+
+func NewLog(path string) *Log {
+	return &Log{
+		path:  path,
+		flush: make(chan int, 1),
+
+		// these options should be overriden by any options in the config
+		LogOutput:              os.Stderr,
+		FlushInterval:          DefaultFlushInterval,
+		SegmentSize:            DefaultSegmentSize,
+		MaxSeriesSize:          DefaultMaxSeriesSize,
+		CompactionThreshold:    DefaultCompactionThreshold,
+		PartitionSizeThreshold: DefaultPartitionSizeThreshold,
+		ReadySeriesSize:        DefaultReadySeriesSize,
+		PartitionCount:         DefaultPartitionCount,
+	}
+}
+
+// Open opens and initializes the Log. Will recover from previous unclosed shutdowns
+func (w *Log) Open() error {
+	// open the partitions
+	w.partitions = make(map[uint8]*partition)
+	for i := uint64(1); i <= w.PartitionCount; i++ {
+		p, err := newPartition(uint8(i), w.path, w.SegmentSize)
+		if err != nil {
+			return err
+		}
+		w.partitions[uint8(i)] = p
+	}
```

Comment:
We aren't going to allow it to change. I think it's more trouble than its worth since we don't flush on startup so we have to be able to support the previous partition scheme. I changed the variable in `Log` to not be exported. Still want the variable there for testing purposes though.